feat!: add `useSetI18nParams` composable

BobbieGoede opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxtjs.com/docs/community/contribution
-->

### 🔗 Linked issue
* #2429 
* #2006
* #1736
<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
* Depends on https://github.com/intlify/routing/pull/82
* Still could add more tests.

This PR adds a composable `useSetI18nParams` which returns a function `setI18nParams` which can be used to set the translated parameters for the current route.

```ts
// `products/[slug].vue` (current page='/en/products/red-mug')
 
const setI18nParams = useSetI18nParams();
setI18nParams({
  en: {
    slug: 'red-mug'
  },
  nl: {
    slug: 'rode-mok'
  }
})
```

The composable function logic is basically a reimplementation of `useLocaleHead` from `vue-i18n-routing` except it uses Nuxt's built in unhead instance to manage and render the tags.

This PR only works combined with https://github.com/intlify/routing/pull/82, as this allows us to change where these dynamic translated parameters are sourced from. ~~In this implementation it uses `useState` to set and retrieve these parameters, which seems to work in my testing.~~

There is likely a better solution for this, but I hope to find a way to resolve the linked issues as I think this is/would be blocking users from moving to v8. If this works, we can always improve the functionality or performance later.

### Goal
I have removed references to using dynamic parameters with `definePageMeta`, the issue with using `definePageMeta` is that it's fixed at build time. 

Workarounds described https://github.com/nuxt-modules/i18n/issues/1736 bypass this by setting `route.meta.nuxtI18n` directly, this works except for updating the SEO tags during SSR, this PR works similar to the workarounds except SEO tags are updated and rendered on the server.

The issue described below also applies to those workarounds.

### Common issue
In the documentation I have added a warning about a common scenario as there is no reactivity during SSR. For example, setting dynamic parameters from within a page will not be reflected in a layout as it would have already been rendered (see [this comment](https://github.com/nuxt/nuxt/issues/18627#issuecomment-1408951142)). ~~This will cause a hydration mismatch if a language switcher is present in a layout or any page or component that renders before setting dynamic parameters.~~ This has been improved to match the workaround, hydration mismatch is avoided as links will be retroactively updated on the client side, this means not all links are rendered correctly on the server-side while the head tags are.

### Remarks
* I haven't removed the logic of dynamic parameters using `definePageMeta` as that would make this a breaking PR. ~~should this be fully removed and/or trigger a deprecation warning when still in use?~~ I have added deprecation warnings which log which pages are using the deprecated method.
* In the future I would like to improve and implement this in `vue-i18n-routing`, as `unhead` can be used there as well. It's not my intention to move code out of that repository, but the current implementation makes use of Nuxt's `unhead` instance.

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

BobbieGoede committed:
feat: add `useSetI18nParams` composable

BobbieGoede committed:
test: test `useI18nParams` composable

BobbieGoede committed:
chore: add `useSetI18nParams` to playground

BobbieGoede committed:
docs: remove and replace dynamic parameter usage with `useSetI18nParams`

BobbieGoede committed:
fix: add `useSetI18nParams` to auto imports

BobbieGoede committed:
test: fix dynamic parameters test

kazupon commented on docs/content/3.options/2.routing.md:
````diff
@@ -126,47 +126,3 @@ Set to a path to which you want to redirect users accessing the root URL (`/`).
 }
 ```
-
-## `dynamicRouteParams`
````
The `dynamicRouteParams` docs should be kept for those who have already implemented it. And a note and link to `useSetI18nParams` should be left for those who have already implemented it.

Let’s remove it in the next minor release, v8.1, after the v8 official release.

kazupon commented on specs/fixtures/issues/2020/nuxt.config.ts:
```diff
@@ -16,7 +16,6 @@ export default defineNuxtConfig({
         iso: 'en-US'
       }
     ],
-    dynamicRouteParams: true,
```
If `dynamicRouteParams` option is used, nuxt i18n should warn to console.

kazupon commented on specs/fixtures/switcher/pages/[...pathMatch].vue:
```diff
@@ -3,11 +3,10 @@ import LangSwitcher from '../components/LangSwitcher.vue'
 
 const route = useRoute()
 
-definePageMeta({
```
if we are using `nuxtI18n` in `definePageMeta`, nuxt i18n should warn the console.

The way to implement it is to implement it as a bundler plugin. The code is passed as a parameter to `transform`, so you can check it at that time.

kazupon reviewed (changes requested):
Great work!
I've just reviewed your PR!
Please check it!

BobbieGoede commented:
Thanks for the feedback! I agree that it's better to keep the docs and refer to the new composable, and add warnings when using the deprecated functionality. 

After more testing with this PR I ran into some SSR/hydration issues that I would like to fix as well. It will take some time to figure out the issues and to make the suggested changes so I'm marking this PR as WIP in the meantime!

BobbieGoede committed:
fix: prevent hydration mismatch by not using `useState`

BobbieGoede committed:
fix: add deprecation warning for `nuxtI18n` usage

BobbieGoede committed:
fix: revert docs and refer to composable

BobbieGoede committed:
docs: improve `useSetI18nParams` documentation

BobbieGoede committed:
test: improve `useSetI18nParams` test

BobbieGoede committed:
chore: cleanup playground

BobbieGoede committed:
fix: invert `useSetI18nParams` defaults

BobbieGoede commented:
While trying to debug the hydration issues I found out I overcomplicated the solution a bit by using `useState`. This feature now internally uses the `nuxtI18n` approach, but improves its usage and ensures the meta tags are updated. I think there are a few ways to improve on the current way we handle head tags, but that will be something I can focus on aftwards.

@kazupon 
I have added the deprecation warnings in the documentation and also added the warning logs when usage is detected, should I add tests to confirm these logs are working as expected? And should it be possible to disable this logs?

BobbieGoede committed:
fix: playground products page not changing locale

kazupon reviewed (approved):
Great works!
LGTM!

kazupon commented:
@BobbieGoede 
> should I add tests to confirm these logs are working as expected? And should it be possible to disable this logs?

The logging test should be added if possible.
As far as I know, warning can be disabled by specifying it in the `useLogger` options.
https://nuxt.com/docs/api/kit/logging

`useLogger` uses consola, I think it may be possible to disable it by setting an ENV or log level.
https://github.com/unjs/consola

BobbieGoede commented:
@kazupon 
>The logging test should be added if possible.

Okay add those soon!

I think I will also change the arguments passed to `useSetI18nParams`, so that instead of `{ addDirAttribute?: boolean, addSeoAttribute?: boolean | SeoAttributesOptions }` it will accept `SeoAttributesOptions | undefined`. Reasoning behind it is that this function shouldn't change the `dir` attribute (will remove that), and right now if `addSeoAttribute` isn't passed the function won't do anything.

BobbieGoede committed:
fix: use `useLogger` for meta deprecation bundler plugin

BobbieGoede committed:
fix: change `useSetI18nParams` parameters

BobbieGoede commented:
@kazupon 
Actually I don't know how to test those logs as they happen at build time 😅

>useLogger uses consola, I think it may be possible to disable it by setting an ENV or log level.
[unjs/consola](https://github.com/unjs/consola)

Not sure if this is possible, as far as I can tell it is possible to set the log level, but changing it would affect all logs, not just for this module.

I did change the composable parameter as I described.

kazupon commented:
@BobbieGoede 

> Actually I don't know how to test those logs as they happen at build time 😅

I think that we can do test at unit test that work unplugin directly. However, that is really difficult as it requires mock.
we can test the `dynamicRouteParams` and `route.meta.nuxtI18n` warning logs at playground. 

I've pull the your branch in my local, and I've checked your branch with playground.
I saw the warning logs.
I can merge your PR, so it looks good for me. :)

We should release rc.9, because that PR might have still issues.
If there is the issue, just fix it!

BobbieGoede commented:
Okay good! I'll merge this and release `rc.9` in a moment 💪

kazupon commented:
@BobbieGoede 
I'll merge your PR!
